### PR TITLE
Allow CompactionSegmentIterator to have custom priority

### PR DESCRIPTION
**Pull request by kfaraz**
### Description

`NewestSegmentFirstIterator` has a lot of code which may be commoned out if we want to implement other types
of compaction search policies. The changes in this PR create a common iterator which may be given any comparator
to prioritize between compactible segments.

### Changes
- Break `NewestSegmentFirstIterator` into two parts
  - `DatasourceCompactibleSegmentIterator` - this contains all of the code originally in `NewestSegmentFirstIterator`
  but now handles a single datasource and allows a priority to be specified
  - `PriorityBasedCompactionSegmentIterator` - contains separate iterator for each datasource and combines the
  results into a single queue to be used by a compaction search policy
- Update `NewestSegmentFirstPolicy` to use the above new classes
- Cleanup `CompactionStatistics` and `AutoCompactionSnapshot`
- Cleanup `CompactSegments`
- Remove unused methods from `Tasks`
- Remove unneeded `TasksTest`
- Move tests from `NewestSegmentFirstIteratorTest` to `CompactionStatusTest`
and `DatasourceCompactibleSegmentIteratorTest`

### Main classes to review
- `DatasourceCompactibleSegmentIterator`
- `CompactSegments`
- `PriorityBasedCompactionSegmentIterator`

### Planned follow up work
- Add other types of segment search policies to compare performance with `NewestSegmentFirstPolicy`.
- Simplify tracking of compaction stats.
- Further cleanup of these classes. This PR has left some unnecessary code blocks in place to allow for a smooth review.

<hr>

This PR has:

- [ ] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

**Review by AmatyaAvadhanula — APPROVED** [review]
Thanks, @kfaraz!

File: server/src/main/java/org/apache/druid/server/coordinator/compact/DatasourceCompactibleSegmentIterator.java [hunk 1]
```diff
@@ -315,27 +291,12 @@ public List<DataSegment> next()
...
-   *
-   * @return An empty {@link SegmentsToCompact} if there are no eligible candidates.
+   * Finds segments to compact together for the given datasource and adds them to
+   * the priority queue.
    */
-  private SegmentsToCompact findSegmentsToCompact(
-      final String dataSourceName,
-      final DataSourceCompactionConfig config
-  )
+  private void findSegmentsToCompact(CompactibleSegmentIterator compactibleSegmentIterator)
```
Comment (gargvishesh) [inline server/src/main/java/org/apache/druid/server/coordinator/compact/DatasourceCompactibleSegmentIterator.java:297]:
```suggestion
  private void findAndEnqueueSegmentsToCompact(CompactibleSegmentIterator compactibleSegmentIterator)
```

Comment (kfaraz) [reply server/src/main/java/org/apache/druid/server/coordinator/compact/DatasourceCompactibleSegmentIterator.java]: Added.

**Comment by kfaraz** [comment]
Thanks for the reviews, @gargvishesh , @AmatyaAvadhanula !

**Pull request merged**